### DISCUSSION TITLE
Support Workflow Commands

### DISCUSSION THREAD

**Discussion by nikeee** [biomejs/biome#441] [Ideas]
(re-post of https://github.com/rome/tools/discussions/4575)

GitHub actions supports putting annotations in source files during a CI run:
https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions

Feature request: `biome ci` should detect if it is running in GitHub actions and emit error/warning messages for linter and style issues.

eslint seems to support this as well. It looks like this (for `no-unused-vars`):
![image](https://github.com/rome/tools/assets/4068864/36433c5e-0bf2-41eb-bacb-860a66e5cbcf)

If there is a reliable way to detect if biome is running in GH Actions[^1], this should be possible to provide without breaking anything. Maybe only consider doing it when using `biome ci`. Eslint seems to do this by default, if it is running in GHA.

[^1]: Should be possible when checking env var [`GITHUB_ACTIONS` for `true`](https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables).

---

Some thoughts on that feature:

I was trying to find out where exactly eslint implements this feature, so it can be implemented in a similar way. After some research, I [found out](https://github.com/sibiraj-s/action-eslint/issues/3#issuecomment-815477091) that it actually is not part of eslint, but of `setup-node`. The action has a couple of application-specific problem matchers:
- eslint: https://github.com/actions/setup-node/blob/main/.github/eslint-compact.json
- eslint again: https://github.com/actions/setup-node/blob/main/.github/eslint-stylish.json
- tsc: https://github.com/actions/setup-node/blob/main/.github/tsc.json

They rely on the output of the respective tools and match them against some regular expressions.

This might only be an option for tools with a very broad adoption, but in principle, this could be done for biome as well. This would effectively "freeze" the output format of biome, since the problem matchers can't be changed that often.

In my opinion, emitting the workflow commands via biome would be easier and more future-proof. However, we have to
1. Detect if we're running in GH Actions (easy)
2. Detect if the output is piped to some other process, as we don't want to print unexpected output to a file. If this is the case, we should not emit these workflow markers. (does GHA emulate a TTY? [Is this possible?](https://github.com/actions/runner/issues/241))
3. If 2. does not have a reliable solution, what to do instead?

As mentioned in https://github.com/rome/tools/discussions/4575, an earlier version of biome had this feature. How was it implemented back then?

**Comment by ematipico** [comment]
Here's an old PR that enabled that feature a long time ago: https://github.com/rome/tools/pull/739

We can use it as an example.

**Reply by nikeee** [reply]
I've drafted a proof-of-concept in a fork:
https://github.com/biomejs/biome/compare/main...nikeee:biome:github-warnings?expand=1

My knowledge of rust is very limited. That's why I could not manage to add the diagnostic description to the output. It also misses a check for `GITHUB_ACTIONS` + passing it down to the CI check.

**Reply by ematipico** [reply]
Would you like to open an issue where we can track the tasks needed to implement this feature?

Then we can follow-up with a PR so that we can help you.

**Reply by ematipico** [reply]
hey @nikeee , would you like to open a PR or an issue? I am very interested in this work

**Reply by nikeee** [reply]
I'll create one

**Reply by nikeee** [reply]
Feature [released in biome 1.5](https://biomejs.dev/blog/biome-v1-5/#the-command-biome-ci-now-prints-diagnostics-in-github-prs). I think we can close this discussion.

**Comment by sairion** [comment]
I was looking for workflow command for Biome, and ended up in @nikeee 's PR, which looks fascinating. Is there any more things need to be done?

**Reply by ematipico** [reply]
That's been implemented. You'll have to use `biome ci` in CI, probably enable the write permissions: 

```yml
permissions:
  pull-requests: write
```
